src/pyjion/intrins.cpp

```diff
@@ -2135,33 +2134,60 @@ PyObject* MethCallN(PyObject* self, PyMethodLocation* method_info, PyObject* arg
             return nullptr;
         }
         auto obj =  method_info->object;
-        auto args_tuple = PyTuple_New(PyTuple_Size(args) + 1);
-
-        assert(obj != nullptr);
-        if (PyTuple_SetItem(args_tuple, 0, obj) == -1){
-            return nullptr;
-        }
-
-        Py_INCREF(obj);
+        if (PyCFunction_Check(target)) {
+            // We allocate an additional two slots. One is for the `self` argument since we're
+            // executing a method. The other is to leave space at the beginning of the vector so we
+            // can use the `PY_VECTORCALL_ARGUMENTS_OFFSET` flag and avoid an allocation in the callee.
+            const auto args_vec_size = PyTuple_Size(args) + 2;
+            auto* args_vec = new PyObject*[args_vec_size];
+            args_vec[1] = obj;
+            Py_INCREF(obj);
+            for (int i = 0; i < PyTuple_Size(args); ++i) {
+                auto* arg = PyTuple_GET_ITEM(args, i);
+                assert(i + 2 < args_vec_size);
+                args_vec[i + 2] = arg;
+                Py_INCREF(arg);
+            }
+#ifdef GIL
+            PyGILState_STATE gstate;
+            gstate = PyGILState_Ensure();
+#endif
+            // The `PY_VECTORCALL_ARGUMENTS_OFFSET` flag lets callees know that they're allowed to
+            // write to `args[-1]` so we should pass the pointer to the first item in our vector and
+            // subtract one from the size argument.
+            res = PyObject_Vectorcall(target, args_vec + 1, (args_vec_size - 1) | PY_VECTORCALL_ARGUMENTS_OFFSET, nullptr);
+#ifdef GIL
+            PyGILState_Release(gstate);
+#endif
+            delete[] args_vec;
+        } else {
+            auto args_tuple = PyTuple_New(PyTuple_Size(args) + 1);
 
-        for (int i = 0 ; i < PyTuple_Size(args) ; i ++){
-            auto ix = PyTuple_GET_ITEM(args, i);
-            assert(ix != nullptr);
-            if (PyTuple_SetItem(args_tuple, i+1, ix) == -1){
+            assert(obj != nullptr);
+            if (PyTuple_SetItem(args_tuple, 0, obj) == -1){
                 return nullptr;
             }
-            Py_INCREF(ix);
-        }
 
+            Py_INCREF(obj);
+
+            for (int i = 0 ; i < PyTuple_Size(args) ; i ++){
+                auto ix = PyTuple_GET_ITEM(args, i);
+                assert(ix != nullptr);
+                if (PyTuple_SetItem(args_tuple, i+1, ix) == -1){
+                    return nullptr;
+                }
+                Py_INCREF(ix);
+            }
 #ifdef GIL
-        PyGILState_STATE gstate;
-        gstate = PyGILState_Ensure();
+            PyGILState_STATE gstate;
+            gstate = PyGILState_Ensure();
 #endif
-        res = PyObject_Call(target, args_tuple, nullptr);
+            res = PyObject_Call(target, args_tuple, nullptr);
 #ifdef GIL
-        PyGILState_Release(gstate);
+            PyGILState_Release(gstate);
 #endif
-        Py_DECREF(args_tuple);
+            Py_DECREF(args_tuple);
```

Comment:
This is from the existing code. But do we need to iterate over the tuple and do `Py_DECREF` on each of the objects since we did `Py_INCREF` as we were adding them to the tuple?

Reply:
This is a bit ugly, but the `args_tuple` comes from `PyTuple_New(n)`, which has a refcnt of 1.
Then the arguments are loaded into that tuple using `emit_tuple_store()`, which adds them without incrementing their references. Then when you decref `args_tuple` on this line it will call `tuple_dealloc` because the refcnt becomes 0. This iterates over the tuple and decrements all items.

Like I said, it's ugly. I could increment the references properly and then iterate them in the functions which use them, but it creates overhead for the sake of readability